Add support for creating certmanager certificates

OlleLarsson opened the pull request:
Example values that i have used:

```
certmanager:
  enabled: true
  
  ca:
    commonName: example.com
    organization:
    - elastisys
  
  elasticsearch:
    transport:
      commonName: nodes.example.com
      organization:
      - elastisys
    admin: 
      commonName: nodes.example.com
      organization:
      - elastisys
```

Commit:
Add support for creating certmanager certificates

ewnetu commented on helm/opendistro-es/values.yaml:
```diff
@@ -11,6 +11,59 @@
+      - client auth
+    transport:
+      enabled: true
+      secretName: es-transport
```
I believe these secret names are going to be used for the different secrets under `elasticsearch:` and `kibana:`. Maybe also add them there.  something like
`elasticsearch:
   ssl:
      transport:
        existingCertSecret: **es-transport**
`?

OlleLarsson replied:
Not really sure i see the problem here. Can't kibana just mount the transport certificate? 
```
kibana
  ssl:
    elasticsearch:
      enabled: true
      existingCertSecret: es-transport
      existingCertSecretCertSubPath: elk-rest-crt.pem
      existingCertSecretKeySubPath: elk-rest-key.pem
      existingCertSecretRootCASubPath: elk-rest-root-ca.pem
```

ewnetu replied:
There is no issue at all. What I was asking is that the secret names are not added there and would be nice to have them. for example currently the secret name for `existingCertSecret` is not specified in values.yaml as you can see below. And I was suggesting to add it (also for the others)
```
kibana
  ssl:
    elasticsearch:
      enabled: false
      existingCertSecret: 
      existingCertSecretCertSubPath: elk-rest-crt.pem
      existingCertSecretKeySubPath: elk-rest-key.pem
      existingCertSecretRootCASubPath: elk-rest-root-ca.pem
```

OlleLarsson replied:
So, you mean that we should not set any default names for the certificates, whereby they must be specified through user passed values?

ewnetu replied:
that or put the secret names in their respective configurations. e.g.,

```
kibana
  ssl:
    elasticsearch:
      enabled: false
      existingCertSecret: es-rest
      existingCertSecretCertSubPath: elk-rest-crt.pem
      existingCertSecretKeySubPath: elk-rest-key.pem
      existingCertSecretRootCASubPath: elk-rest-root-ca.pem
```

lentzi90 reviewed (commented):
I think this looks good, but I haven't tested it.
One question that popped up while reviewing:
Why do we need a certificate for kibana in this way? Can we not use the normal letsencrypt certificate with Ingress?

OlleLarsson commented:
> I think this looks good, but I haven't tested it.
> One question that popped up while reviewing:
> Why do we need a certificate for kibana in this way? Can we not use the normal letsencrypt certificate with Ingress?

Well, i guess the traffic is not encrypted between nginx and kibana, which i am not sure is a problem or not.

lentzi90 commented:
> > I think this looks good, but I haven't tested it.
> > One question that popped up while reviewing:
> > Why do we need a certificate for kibana in this way? Can we not use the normal letsencrypt certificate with Ingress?
> 
> Well, i guess the traffic is not encrypted between nginx and kibana, which i am not sure is a problem or not.

For compliance, we probably want it at some point, but we are not doing it with any other services at the moment. I'd say skip it for now. When we get to the point when we want mTLS, let's install a service mesh and enable it for all services at once. :slightly_smiling_face:

OlleLarsson commented:
> For compliance, we probably want it at some point, but we are not doing it with any other services at the moment. I'd say skip it for now. When we get to the point when we want mTLS, let's install a service mesh and enable it for all services at once.

Sure, sound resonable. Let's say that i do a port-forward to kibana, then the traffic will not be encrypted if we remove tls from kibana, i guess?

OlleLarsson commented:
Should we then also not use tls on the rest layer of elastisearch?

lentzi90 commented:
> Sure, sound resonable. Let's say that i do a port-forward to kibana, then the traffic will not be encrypted if we remove tls from kibana, i guess?

Yes that is correct.

> Should we then also not use tls on the rest layer of elastisearch?

If we can (and do) expose it through an Ingress I would say we should do TLS in the ingress controller. I guess this is what we currently do but I'm not certain.

OlleLarsson commented:
> > Sure, sound resonable. Let's say that i do a port-forward to kibana, then the traffic will not be encrypted if we remove tls from kibana, i guess?
> 
> Yes that is correct.
> 
> > Should we then also not use tls on the rest layer of elastisearch?
> 
> If we can (and do) expose it through an Ingress I would say we should do TLS in the ingress controller. I guess this is what we currently do but I'm not certain.

Ok. Should i disable and leave the functionality in or should i remove it completely?
Edit: i'll just remove it alltogether

OlleLarsson commented:
I have updated the pr now. I have removed the rest and kibana certificates.
Now the secrets name will be taken from the `existingCertSecret` in the ssl configuration for elasticsearch. No default names are provided.
```
secretName: {{ .Values.elasticsearch.ssl.transport.existingCertSecret }}
secretName: {{ .Values.elasticsearch.ssl.admin.existingCertSecret }}
```

ewnetu reviewed (approved):
Looks good!
